### PR TITLE
feat: Add schedule-or-reschedule method to the SchedulerClient

### PR DESCRIPTION
**Pull request by beilCrxmarkets**
## feat: Add rescheduleOrCreate method to the SchedulerClient
When invoking SchedulerClient.reschedule() with a non-existent TaskInstance, an exception will be raised. Therefore, it's essential to verify the existence of a scheduled TaskInstance prior to attempting a reschedule. To streamline this process, we propose the introduction of a scheduleOrCreate method. This method would automatically schedule a TaskInstance if none exists, making it particularly beneficial for dynamically recurring tasks that require initial setup.

We have given considerable thought to finding an appropriate name for this method, but unfortunately, we haven't arrived at a satisfactory conclusion. The verb "schedule" best describes its functionality; however, it cannot be used as there is already a method with that name. Since "rescheduleOrSchedule" sounds awkward, we have settled on "rescheduleOrCreate" for now. We are open to considering other names for this method.

## Fixes
[#539](https://github.com/kagkarlsson/db-scheduler/issues/539)

## Reminders
- [x] Added/ran automated tests
- [x] Update README and/or examples
- [x] Ran `mvn spotless:apply`

---
cc @kagkarlsson

### CODE REVIEW

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 1]
```diff
@@ -352,6 +380,26 @@ public <T> void reschedule(TaskInstanceId taskInstanceId, Instant newExecutionTi
       }
     }
 
+    @Override
+    public <T> void rescheduleOrCreate(TaskInstance<T> taskInstance, Instant executionTime) {
+      Optional<ScheduledExecution<Object>> executionObject =
+          this.getScheduledExecutions().stream()
+              .filter(e -> e.getTaskInstance().getId().equals(taskInstance.getId()))
+              .findFirst();
```
Comment (kagkarlsson) [inline db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java:388]: We should avoid fetch all here, instead use
```
Optional<ScheduledExecution<Object>> getScheduledExecution(TaskInstanceId taskInstanceId);
```

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 2]
```diff
@@ -352,6 +380,26 @@ public <T> void reschedule(TaskInstanceId taskInstanceId, Instant newExecutionTi
...
+      if (executionObject.isPresent()) {
+        this.reschedule(taskInstance, executionTime, taskInstance.getData());
+      } else {
+        this.scheduleIfNotExists(taskInstance, executionTime);
+      }
```
Comment (kagkarlsson) [inline db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java:393]: As long as we do not lock the row, there will be the possibility of race-conditions. 

* Maybe log a warn if `scheduleIfNotExists` returns false
* Also, maybe make `reschedule` return the success-code and warn on false on that as well.

Comment (kagkarlsson) [inline db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java:393]: As long as we do not lock the row, there will be the possibility of race-conditions. 

* Maybe log a warn if `scheduleIfNotExists` returns false
* Also, maybe make `reschedule` return the success-code and warn on false on that as well.

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 3]
```diff
@@ -352,6 +380,26 @@ public <T> void reschedule(TaskInstanceId taskInstanceId, Instant newExecutionTi
...
+    }
+
+    @Override
+    public <T> void rescheduleOrCreate(SchedulableInstance<T> schedulableInstance) {
+      rescheduleOrCreate(
+          schedulableInstance.getTaskInstance(),
+          schedulableInstance.getNextExecutionTime(clock.now()));
+    }
```
Comment (kagkarlsson) [inline db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java:401]: Agreed that it is hard to come up with a good name for this operation.. 🤔 

rescheduleOrCreate
scheduleOrReschedule
scheduleOrMove
scheduleOrReplace
schedule(id, WhenExists.RESCHEDULE).  (OVERWRITE/DO_NOTHING...)
..

What do you think about a variant of the last one, where you have a parameter for what to do when the task-instance already exists?

**Review by kagkarlsson — COMMENTED** [review]
Thank you for contributing! A couple of comments :)

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java (hunk 1 above)
Comment (beilCrxmarkets) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: You are right about that. I have adopted your suggestion.

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java (hunk 2 above)
Comment (beilCrxmarkets) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: In my opinion, scheduleIfNotExists should generally create a warning log entry in the event of failure. This is already the case. I therefore do not see the need to generate an additional log entry for this case. 

reschedule now also returns a boolean value and also creates a warning log entry in the event of an error. The only question here is how useful this is at the moment, as an ExecutionException is thrown in the only possible error case. It is therefore currently not possible for reschedule to return false, as an exception is thrown in this case.

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java (hunk 3 above)
Comment (beilCrxmarkets) [reply db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: schedule as the method name with an additional parameter seems most intuitive to me. 

I have therefore adapted the code accordingly. 
schedule now also returns a boolean value, just like scheduleIfNotExists and reschedule.

File: db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 4]
```diff
@@ -322,42 +365,77 @@ public <T> void schedule(SchedulableInstance<T> schedulableInstance) {
...
       String instanceId = taskInstanceId.getId();
-      Optional<Execution> execution = taskRepository.getExecution(taskName, instanceId);
-      if (execution.isPresent()) {
-        if (execution.get().isPicked()) {
-          throw new TaskInstanceCurrentlyExecutingException(taskName, instanceId);
-        }
+      Execution execution =
+          taskRepository
+              .getExecution(taskName, instanceId)
+              .orElseThrow(() -> new TaskInstanceNotFoundException(taskName, instanceId));
```
Comment (beilCrxmarkets) [inline db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java:420]: With the use of `orElseThrow` we can get rid of the big brackets which goes around 10+ lines of code, which should improve the code readability.

**Review by kagkarlsson — APPROVED** [review]
Looks good 👍

**Pull request merged**